src/main/java/dev/koifysh/archipelago/Client.java

```diff
@@ -117,45 +141,87 @@ public void removeTag(String tag) {
+                    if(localGamesList.containsKey(gameName) && localGamesList.get(gameName).isDirectory()){
+                        //check all checksums
+                        for(File version : localGamesList.get(gameName).listFiles()){
+                            if(versions.containsKey(gameName) && versions.get(gameName).equals(version.getName())){
+                                try(FileReader reader = new FileReader(version)){
+                                    updateDataPackage(gson.fromJson(reader, DataPackage.class));
+                                    LOGGER.info("Read datapackage for Game: ".concat(gameName).concat(" Checksum: ").concat(version.getName()));
+                                } catch (IOException e){
+                                    LOGGER.info("Failed to read a datapackage. Creating a new one.");
+                                    dataPackage = new DataPackage();
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        } else{
+            //cache doesn't exist. Create the filepath
+            boolean success = directoryPath.mkdirs();
+            if(success){
+                //make a new datapackage
+                LOGGER.info("DataPackage directory didn't exist. Starting from fresh.");
+                dataPackage = new DataPackage();
+                //saveDataPackage();
+            } else{
+                //failed to make directories?
+                LOGGER.severe("Failed to make directories for datapackage cache.");
+            }
         }
     }
 
-    void saveDataPackage() {
-        try {
-            File dataPackageFile = new File(dataPackageLocation);
-
-            //noinspection ResultOfMethodCallIgnored
-            dataPackageFile.getParentFile().mkdirs();
-            //noinspection ResultOfMethodCallIgnored
-            dataPackageFile.createNewFile();
-
-            FileOutputStream fileOut = new FileOutputStream(dataPackageFile);
-            ObjectOutputStream objectOut = new ObjectOutputStream(fileOut);
-
-            objectOut.writeObject(dataPackage);
-
-
-            fileOut.close();
-            objectOut.close();
+    public void saveDataPackage() {
+        
+        //Loop through games to ensure we have folders for each of them in the cache
+        for(String gameName : games){
+            File gameFolder = new File(dataPackageLocation.concat(gameName));
+            if(!gameFolder.exists()){
+                //game folder not found. Make it
+                gameFolder.mkdirs();
+            }
 
-        } catch (IOException e) {
-            LOGGER.warning("unable to save DataPackage.");
+            //save the datapackage
+            for(Map.Entry<String, String> game : versions.entrySet()){
+                //if game is not in list of games
+                if (!games.contains(game.getKey()))
+                    continue;
```

Comment:
Would a simple lookup here be fine?  Especially since we're in the context of looping over `games`?

I.e.

```java
  String gameVersion = versions.get(gameName);
  if(gameVersion == null) {
    continue;
  }
  
```